File: stasis-core/src/de/boxxit/statis/serializer/SerializableExceptionSerializer.java
package de.boxxit.statis.serializer;

import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.Serializer;
import com.esotericsoftware.kryo.io.Input;
import com.esotericsoftware.kryo.io.Output;
import de.boxxit.statis.SerializableException;

/**
 * User: Christian Fruth
 */
public class SerializableExceptionSerializer extends Serializer<SerializableException>
{
	public SerializableExceptionSerializer()
	{
	}

	@Override
	public void write(Kryo kryo, Output output, SerializableException object)
	{
		kryo.writeObjectOrNull(output, object.getId(), String.class);
		kryo.writeObjectOrNull(output, object.getMessage(), String.class);
	}

	@Override
	public SerializableException read(Kryo kryo, Input input, Class<SerializableException> type)
	{
		String id = kryo.readObjectOrNull(input, String.class);
		String message = kryo.readObjectOrNull(input, String.class);
		return new SerializableException(id, message);
	}
}

File: stasis-ios/Stasis/SerializableExceptionSerializer.m
//
//  SerializableExceptionSerializer.m
//  Stasis
//
//  Created by Christian Fruth on 06.03.13.
//  Copyright (c) 2013 Boxx IT Solutions e.K. All rights reserved.
//

#import "SerializableExceptionSerializer.h"
#import "SerializableException.h"
#import "Kryo.h"

@implementation SerializableExceptionSerializer

- (BOOL)acceptsNull
{
	return NO;
}

- (void)write:(Kryo *)kryo value:(id)value to:(KryoOutput *)output
{
	[NSException raise:NSInternalInconsistencyException format:@"%@ is an unsupported operation", NSStringFromSelector(_cmd)];
}

- (id)read:(Kryo *)kryo withClass:(Class)clazz from:(KryoInput *)input
{
	NSString *type = [input readString];
	NSString *message = [input readString];
	SInt32 stackLength = [input readIntOptimizePositive:YES];
	NSMutableArray *stackSymbols = [NSMutableArray arrayWithCapacity:stackLength];
	
	for (SInt32 i = 0; i < stackLength; ++i)
	{
		NSString *className = [input readString];
		NSString *methodName = [input readString];
		NSString *fileName = [input readString];
		SInt32 lineNumber = [input readIntOptimizePositive:YES];
		NSString *stackSymbol = [NSString stringWithFormat:@"%@.%@(%@:%d)", className, methodName, fileName, (int)lineNumber];
		
		[stackSymbols addObject:stackSymbol];
	}

	return [SerializableException exceptionWithType:type message:message andStackSymbols:stackSymbols];
}

@end
